controllers/cruisecontroloperation_controller.go

```diff
@@ -0,0 +1,473 @@
+			},
+		})
+
+	return builder
+}
+
+func isFinalizeNeeded(operation *banzaiv1alpha1.CruiseControlOperation) bool {
+	return controllerutil.ContainsFinalizer(operation, cruiseControlOperationFinalizer) && !operation.ObjectMeta.DeletionTimestamp.IsZero()
+}
+
+func getKafkaClusterReference(operation *banzaiv1alpha1.CruiseControlOperation) (client.ObjectKey, error) {
+	if operation.GetClusterRef() == "" {
+		return client.ObjectKey{}, errors.NewWithDetails("could not find kafka cluster reference label for CruiseControlOperation", "missing label", banzaiv1beta1.KafkaCRLabelKey, "name", operation.GetName(), "namespace", operation.GetNamespace())
+	}
+	return client.ObjectKey{
+		Name:      operation.GetClusterRef(),
+		Namespace: operation.GetNamespace(),
+	}, nil
+}
+
+func updateResult(log logr.Logger, res *scale.Result, operation *banzaiv1alpha1.CruiseControlOperation, isAfterExecution bool) error {
+	// This can happen rarely when the max cached completed user tasks reached
+	if res == nil {
+		log.Error(missingCCResErr, "Cruise Control's max.cached.completed.user.tasks configuration value probably too small. Missing user task state is handled as completedWithError", "name", operation.GetName(), "namespace", operation.GetNamespace(), "task ID", operation.GetCurrentTaskID())
+		res = &scale.Result{
+			TaskID: operation.GetCurrentTaskID(),
+			State:  banzaiv1beta1.CruiseControlTaskCompletedWithError,
+			Err:    missingCCResErr.Error(),
+		}
+	}
+
+	operation.Status.ErrorPolicy = operation.Spec.ErrorPolicy
+	task := operation.GetCurrentTask()
+
+	if (task.State == banzaiv1beta1.CruiseControlTaskCompleted || task.State == banzaiv1beta1.CruiseControlTaskCompletedWithError) && task.Finished == nil {
+		task.Finished = &v1.Time{Time: time.Now()}
+	}
+
+	// Add the failed task into the status.failedTasks slice only when the update is happened after executing the task
+	if isAfterExecution && task.Finished != nil && task.State == banzaiv1beta1.CruiseControlTaskCompletedWithError {
+		if len(operation.Status.FailedTasks) >= defaultFailedTasksHistoryMaxLength {
+			operation.Status.FailedTasks = append(operation.Status.FailedTasks[1:], *task)
+		} else {
+			operation.Status.FailedTasks = append(operation.Status.FailedTasks, *task)
+		}
+		operation.Status.NumberOfRetries += 1
+		task.SetDefaults()
+	}
+
+	if isAfterExecution {
+		if task.Started == nil {
+			startTime, err := time.Parse(time.RFC1123, res.StartedAt)
+			if err != nil {
+				return errors.WrapIff(err, "could not parse user task start time from Cruise Control API")
+			}
+			task.Started = &v1.Time{Time: startTime}
+		}
+		task.ID = res.TaskID
+		task.Summary = parseSummary(res.Result)
+		task.ErrorMessage = res.Err
+		task.HTTPRequest = res.RequestURL
+		task.HTTPResponseCode = &res.ResponseStatusCode
+	}
+
+	task.State = res.State
+
+	return nil
+}
+
+// updateActiveTasks updates the state of the CruiseControlOperation from the CruiseControlTasksAndStates instance by getting their
```

Comment:
Looks like `updateActiveTasks` is the old function name and it does't match the latest function name
```suggestion
// updateCurrentTasks updates the state of the CruiseControlOperation from the CruiseControlTasksAndStates instance by getting their
```